pennylane/qnodes/base.py

```diff
@@ -264,6 +266,55 @@ def qfunc(a, w):
             "The QNode can only be drawn after its CircuitGraph has been constructed."
         )
 
+    def set_trainable_args(self, arg_indices):
+        """Store the indices of quantum function positional arguments
+        that support differentiability.
+
+        Args:
+            args (None or Set[int]): Differentiable positional argument indices. A
+                value of ``None`` means that all argument indices are differentiable.
+        """
+        if not self.mutable and self.circuit is not None:
+
+            if self.get_trainable_args() == arg_indices:
+                return
+
+            raise QuantumFunctionError(
+                "The trainability of arguments on immutable QNodes cannot be modified after the first evaluation."
+            )
+
+        if not arg_indices:
```

Comment:
There is some ambiguity here, since both `None` and `set()` would lead to this being `True` (as well as `[]`). Based on the docstring above, `None` means that all argument indices are trainable

Reply:
Oh good catch. I've separated this out into two checks:
```python

if arg_indices is None:
    # all arguments are differentiable
    self._trainable_args = None
    return

if not arg_indices:
    # The provided arg_indices are an empty serquence;
    # no arguments are differentiable.
    self._trainable_args = set()
    return
```
and will add a test for the case of no differentiable arguments.